python/samples.cpp

```diff
@@ -0,0 +1,155 @@
+struct sampler_state {
+    std::mutex mutex;
+    std::unordered_map<arb::cell_member_type, std::vector<trace_entry>> buffer;
+
+    std::vector<trace_entry>& locked_sampler_vec(arb::cell_member_type pid) {
+        // lock the mutex, s.t. other threads cannot write
+        std::lock_guard<std::mutex> lock(mutex);
+        // return or create entry
+        return buffer[pid];
+    }
+
+    // helper function to search probe id in buffer
+    bool has_pid(arb::cell_member_type pid) {
+        std::unordered_map<arb::cell_member_type, std::vector<trace_entry>>::iterator it = buffer.find(pid);
+        return (it != buffer.end());
+    }
+
+    // helper function to push back to locked vector
+    void push_back(arb::cell_member_type pid, trace_entry value) {
+        auto& v = locked_sampler_vec(pid);
+        v.push_back(std::move(value));
+    }
+
+    // helper function to return whole buffer
+    const std::unordered_map<arb::cell_member_type, std::vector<trace_entry>>& samples() const {
+        return buffer;
+    }
+
+    // helper function to return trace entry of probe id
+    const std::vector<trace_entry>& samples_of(arb::cell_member_type pid) {
+        return buffer[pid];
+    }
+};
+
+// A functor that models arb::sampler_function.
+// Holds a shared pointer to the trace_entry used to store the samples, so that if
+// the trace_entry in sample_recorder is garbage collected in Python, stores will
+// not seg fault.
+
+struct sample_callback {
+    std::shared_ptr<sampler_state> sample_store;
+
+    sample_callback(const std::shared_ptr<sampler_state>& state):
+        sample_store(state)
+    {}
+
+    void operator() (arb::cell_member_type probe_id, arb::probe_tag tag, std::size_t n, const arb::sample_record* recs) {
+        // lock before write
+        auto& v = sample_store->locked_sampler_vec(probe_id);
+        for (std::size_t i = 0; i<n; ++i) {
+            if (auto p = arb::util::any_cast<const double*>(recs[i].data)) {
+                v.push_back({recs[i].time, *p});
+            }
+            else {
+                throw std::runtime_error("unexpected sample type");
+            }
+        }
+    };
+};
+
+// Helper type for recording samples from a simulation.
+// This type is wrapped in Python, to expose sample_recorder::sample_store.
+struct sample_recorder {
+    std::shared_ptr<sampler_state> sample_store;
+
+    sample_callback callback() {
+        // initialize the sample_store
+        sample_store = std::make_shared<sampler_state>();
+
+        // The callback holds a copy of sample_store, i.e. the shared
+        // pointer is held by both the sample_recorder and the callback, so if
+        // the sample_recorder is destructed in the calling Python code, attempts
+        // to write to sample_store inside the callback will not seg fault.
+        return sample_callback(sample_store);
+    }
+
+    const std::vector<trace_entry>& samples(arb::cell_member_type pid) const {
```

Comment:
Put `return` at the end of the function, and handle the error case as a special case.
```C++
if (!sample_store->has_pid(pid))
    throw ...;
return sample_store->probe_buffer(pid);
```

Reply:
Done.